File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/PolylineOverlayHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.Overlay
import com.naver.maps.map.overlay.PolylineOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler.Companion.getterName
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPolylineOverlay
import io.flutter.plugin.common.MethodChannel

internal interface PolylineOverlayHandler : OverlayHandler {
    fun handlePolylineOverlay(
        polylineOverlay: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (polylineOverlay as PolylineOverlay).let { p ->
        when (method) {
            NPolylineOverlay.coordsName -> setCoords(p, arg!!)
            NPolylineOverlay.colorName -> setColor(p, arg!!)
            NPolylineOverlay.widthName -> setWidth(p, arg!!)
            NPolylineOverlay.lineCapName -> setLineCap(p, arg!!)
            NPolylineOverlay.lineJoinName -> setLineJoin(p, arg!!)
            NPolylineOverlay.patternName -> setPattern(p, arg!!)
            getterName(NPolylineOverlay.boundsName) -> getBounds(p, result::success)
            else -> result.notImplemented()
        }
    }

    fun setCoords(polylineOverlay: PolylineOverlay, rawCoords: Any)

    fun setColor(polylineOverlay: PolylineOverlay, rawColor: Any)

    fun setWidth(polylineOverlay: PolylineOverlay, rawWidthDp: Any)

    fun setLineCap(polylineOverlay: PolylineOverlay, rawLineCap: Any)

    fun setLineJoin(polylineOverlay: PolylineOverlay, rawLineJoin: Any)

    fun setPattern(polylineOverlay: PolylineOverlay, patternDpList: Any)

    fun getBounds(polylineOverlay: PolylineOverlay, success: (bounds: Map<String, Any>) -> Unit)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/PolylineOverlayHandler.swift
import NMapsMap
import Flutter

internal protocol PolylineOverlayHandler: OverlayHandler {
    func setCoords(_ polylineOverlay: NMFPolylineOverlay, rawCoords: Any)

    func setColor(_ polylineOverlay: NMFPolylineOverlay, rawColor: Any)

    func setWidth(_ polylineOverlay: NMFPolylineOverlay, rawWidth: Any)

    func setLineCap(_ polylineOverlay: NMFPolylineOverlay, rawLineCap: Any)

    func setLineJoin(_ polylineOverlay: NMFPolylineOverlay, rawLineJoin: Any)

    func setPattern(_ polylineOverlay: NMFPolylineOverlay, patternList: Any)

    func getBounds(_ polylineOverlay: NMFPolylineOverlay, success: (_ bounds: Dictionary<String, Any>) -> Void)
}

internal extension  PolylineOverlayHandler {
    func handlePolylineOverlay(polylineOverlay: NMFOverlay,
                               method: String,
                               args: Any?,
                               result: @escaping FlutterResult) {
        let polylineOverlay = polylineOverlay as! NMFPolylineOverlay
        switch method {
        case NPolylineOverlay.coordsName: setCoords(polylineOverlay, rawCoords: args!)
        case NPolylineOverlay.colorName: setColor(polylineOverlay, rawColor: args!)
        case NPolylineOverlay.widthName: setWidth(polylineOverlay, rawWidth: args!)
        case NPolylineOverlay.lineCapName: setLineCap(polylineOverlay, rawLineCap: args!)
        case NPolylineOverlay.lineJoinName: setLineJoin(polylineOverlay, rawLineJoin: args!)
        case NPolylineOverlay.patternName: setPattern(polylineOverlay, patternList: args!)
        case getterName(NPolylineOverlay.boundsName): getBounds(polylineOverlay, success: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
